v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/bigtablechangestreamstopubsub/FailsafeModJsonToPubsubMessageTransformer.java

```diff
@@ -0,0 +1,311 @@
+              publishModJsonStringToPubsubMessage(failsafeModJsonString.getPayload());
+
+          context.output(pubSubMessage);
+        } catch (Exception e) {
+          LOG.error("Writing to transform dead letter out", e);
+          context.output(
+              transformDeadLetterOut,
+              FailsafeElement.of(failsafeModJsonString)
+                  .setErrorMessage(e.getMessage())
+                  .setStacktrace(Throwables.getStackTraceAsString(e)));
+        }
+      }
+
+      /* Schema Details:  */
+      private PubsubMessage publishModJsonStringToPubsubMessage(String modJsonString)
+          throws Exception {
+        String messageFormat = pubSubUtils.getDestination().getTopicMessageFormat();
+        String messageEncoding = pubSubUtils.getDestination().getTopicMessageEncoding();
+        Publisher publisher = null;
+        ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
+
+
+        block:
+
+        switch (messageFormat) {
+          case "AVRO":
+            ChangelogEntry changelogEntry = mapJsonToChangelogEntry(modJsonString);
+            Encoder encoder = null;
+            switch (messageEncoding) {
+              case "BINARY":
+                encoder = EncoderFactory.get().directBinaryEncoder(byteStream, /* reuse= */ null);
+                break;
+              case "JSON":
+                encoder =
+                    EncoderFactory.get().jsonEncoder(ChangelogEntry.getClassSchema(), byteStream);
+                break;
+              default:
+                break block;
+            }
+            changelogEntry.customEncode(encoder);
+            encoder.flush();
+
+            // Publish the encoded object as a Pub/Sub message.
+            ByteString data = ByteString.copyFrom(byteStream.toByteArray());
+            PubsubMessage message = PubsubMessage.newBuilder().setData(data).build();
+
+            ApiFuture<String> future = publisher.publish(message);
+            break;
+
+          case "Protocol Buffer":
+            ChangelogEntryProto changelogEntryProto = mapJsonToChangelogEntryProto(modJsonString);
+            data = ByteString.copyFrom(byteStream.toByteArray());
+            PubsubMessage.Builder protoMessage = PubsubMessage.newBuilder();
+            switch (messageEncoding) {
+              case "BINARY":
+                protoMessage.setData(changelogEntryProto.toByteString());
+                break;
+              case "JSON":
+                String jsonString =
+                    JsonFormat.printer()
+                        .omittingInsignificantWhitespace()
+                        .print(changelogEntryProto);
+                protoMessage.setData(ByteString.copyFromUtf8(jsonString));
+                break;
+              default:
+                break block;
+            }
+              future = publisher.publish(protoMessage.build());
+              break;
+
+          case "JSON":
+            byte[] encodedRecords = modJsonString.getBytes();
```

Comment:
I think this is taking an internal representation (Mod) and publishing it as is. I don't think this is the right way to do it. There is also a philosophical question how byte array values should be represented in JSON

Reply:
I am addressing this in my PR for CDC->GCS which I am planning to publish today, something you might want to replicate here